test/asn1_time_test.c

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright 1999-2016 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Time tests for the asn1 module */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <openssl/asn1.h>
+#include <openssl/evp.h>
+#include <openssl/objects.h>
+#include "test_main.h"
+#include "testutil.h"
+#include "e_os.h"
+
+struct testdata {
+    char *data;  /* TIME string value */
+    int type;    /* GENERALIZED OR UTC */
+    int check;   /* check result */
+    time_t t;    /* expected */
+    int cmp_result;
+    int convert;
+    size_t time_t_size_req;
+};
+
+/* ASSUMES SIGNED TIME_T */
```

Comment:
I failed to clarify that "problem actually goes even beyond 32 bits" means that test fails on pair of 32-bit platforms readily available to me, linux-x86 and linux-armv4. But I haven't actually dug into code...